rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java

```diff
@@ -514,4 +539,82 @@ else if (Character.isSupplementaryCodePoint(currentChar)) {
+    // (i.e., semantic token types).
+    private interface CategoryPatch extends BiFunction<IConstructor, String, String> {};
+
+    private static class DefaultCategoryPatch implements CategoryPatch {
+        @Override
+        public String apply(IConstructor prod, String defaultCategory) {
+            return defaultCategory;
+        }
+    }
+
+    private static class RascalCategoryPatch implements CategoryPatch {
+
+        // As an optimization, productions that have already been mapped in the
+        // past, are cached for fast lookups in the future. There are separate
+        // caches for productions that *do* need to be mapped to a new category
+        // and those that *do not* need this.
+        private Cache<IConstructor, String>  doPatch = Caches.build();
+        private Cache<IConstructor, Boolean> doNotPatch = Caches.build();
+
+        @SuppressWarnings("java:S131") // Switches without defaults are intended in this method
+        @Override
+        public String apply(IConstructor prod, String defaultCategory) {
+
+            // Check the caches
+            var category = doPatch.getIfPresent(prod);
+            if (category != null) {
+                return category;
+            }
+            if (doNotPatch.getIfPresent(prod) != null) {
+                return defaultCategory; // Possibly different each time this cache is hit
+            }
+
+            // Apply the patch
+            var def = ProductionAdapter.getDefined(prod);
+
+            if (isLabeledLiteral(def)) {
+                switch (SymbolAdapter.getLabel(def)) {
+                case "integer":
+                case "real":
+                case "rational":
+                    return Caches.putThenGet(doPatch, prod, SemanticTokenTypes.Number);
+                case "location":
+                    return Caches.putThenGet(doPatch, prod, SemanticTokenTypes.String);
+                case "regExp":
+                    return Caches.putThenGet(doPatch, prod, SemanticTokenTypes.Regexp);
+                }
+            }
+
+            doNotPatch.put(prod, Boolean.TRUE);
+            return defaultCategory;
+        }
+
+        private boolean isLabeledLiteral(IConstructor def) {
+            return SymbolAdapter.isLabel(def) &&
+                SymbolAdapter.getName(SymbolAdapter.getLabeledSymbol(def)).equals("Literal");
+        }
+    }
+
+    private static class Caches {
+
+        public static <K, V> Cache<K, V> build() {
+            // The usage of weak keys and values ensures: (1) identity equality
+            // is used to compare keys; (2) entries are automatically evicted
+            // and garbage-collected when they are no longer strongly reachable.
+            return Caffeine.newBuilder()
+                .weakKeys()
+                .weakValues()
+                .build();
+        }
+
+        private static <K, V> V putThenGet(Cache<K, V> cache, K k, V v) {
```

Comment:
in a way, the `Cache::get` function of caffeine does this already. at the cost of a mapping function.

Reply:
Fixed. (The cost of the mapping function should be negligible in this case.)

Reply:
Agreed